seatunnel-connectors-v2/connector-iceberg/src/main/java/org/apache/seatunnel/connectors/seatunnel/iceberg/compaction/IcebergCompactionHandler.java

```diff
@@ -0,0 +1,220 @@
+import org.apache.seatunnel.api.table.type.SeaTunnelRow;
+import org.apache.seatunnel.connectors.seatunnel.iceberg.IcebergTableLoader;
+import org.apache.seatunnel.connectors.seatunnel.iceberg.sink.writer.RecordWriter;
+import org.apache.seatunnel.connectors.seatunnel.iceberg.sink.writer.WriteResult;
+import org.apache.seatunnel.connectors.seatunnel.iceberg.source.split.IcebergFileScanTaskSplit;
+
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.RewriteFiles;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.CommitStateUnknownException;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.util.Tasks;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class IcebergCompactionHandler {
+
+    public static final String COMPACTION_ACTION_MESSAGE = "compaction_end";
+    public static final String CURRENT_SNAPSHOT_ID = "current_snapshot_id";
+    public static final String DELETED_FILES = "deleted_files";
+
+    private final Table table;
+    private final FileIO fileIO;
+    private final RecordWriter writer;
+
+    public IcebergCompactionHandler(IcebergTableLoader icebergTableLoader, RecordWriter writer) {
+        this.table = icebergTableLoader.loadTable();
+        this.fileIO = this.table.io();
+        this.writer = writer;
+    }
+
+    public static void emitCompactionEndRecord(
+            Collector<SeaTunnelRow> output, List<IcebergFileScanTaskSplit> finishedSplits) {
+        if (finishedSplits.isEmpty()) {
+            log.warn("No compaction split, skip it.");
+            return;
+        }
+        List<String> deletedFiles =
+                finishedSplits.stream()
+                        .map(split -> split.getTask().file().path().toString())
+                        .collect(Collectors.toList());
+
+        Optional<IcebergFileScanTaskSplit> firstSplit = finishedSplits.stream().findFirst();
+        if (!firstSplit.isPresent()) {
+            log.warn("Failed to find any splits, skipping compaction end record emission.");
+            return;
+        }
+        IcebergFileScanTaskSplit oneSplit = firstSplit.get();
+        SeaTunnelRow compactionEndRecord = new SeaTunnelRow(0);
+        compactionEndRecord.setTableId(oneSplit.getTablePath().getFullName());
+
+        Map<String, Object> options = new HashMap<>();
+        options.put(COMPACTION_ACTION_MESSAGE, COMPACTION_ACTION_MESSAGE);
+        options.put(CURRENT_SNAPSHOT_ID, oneSplit.getCurrentSnapshotId());
+        options.put(DELETED_FILES, deletedFiles);
+        compactionEndRecord.setOptions(options);
+        output.collect(compactionEndRecord);
```

Comment:
So only when source and sink both are iceberg then SeaTunnel can support file compaction?

Reply:
Open a independent task to compaction the iceberg table and reuse the logic of the source

Reply:
why not do it only in sink? We can read data and do file compaction in sink, it's more useful.

Reply:
cc @hailin0

Reply:
> why not do it only in sink? We can read data and do file compaction in sink, it's more useful.

It's not appropriate to perform compression during the task execution.

Reply:
> why not do it only in sink? We can read data and do file compaction in sink, it's more useful.

It's not appropriate to perform compaction during the task execution.

Reply:
We can add a switch configuration, allowing users to choose whether to use it.

Reply:
> @sunxiaojian We can add a switch configuration, allowing users to choose whether to use it.

https://github.com/apache/iceberg/blob/main/flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/actions/RewriteDataFilesAction.java